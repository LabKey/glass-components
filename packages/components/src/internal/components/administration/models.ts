interface Member {
    id: number;
    name: string;
    type: string;
}

interface Group {
    groupName: string;
    members: Member[];
}

export interface GroupMembership {
    [key: string]: Group;
}
