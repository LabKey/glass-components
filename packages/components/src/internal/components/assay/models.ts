/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { immerable, produce } from 'immer';

import { Operation } from '../../../public/QueryColumn';

import { AssayDefinitionModel } from '../../AssayDefinitionModel';
import { LoadingState } from '../../../public/LoadingState';

import { SelectInputChange } from '../forms/input/SelectInput';

import { AssayWizardModel } from './AssayWizardModel';

export interface AssayPropertiesPanelProps {
    operation: Operation;
    model: AssayWizardModel;
    onChange: (fieldValues: any, isChanged?: boolean) => void;
    onWorkflowTaskChange?: SelectInputChange;
}

export class AssayUploadResultModel {
    [immerable] = true;

    readonly assayId: number;
    readonly batchId: number;
    readonly runId: number;
    readonly success: boolean;
    readonly successurl?: string;
    readonly jobId?: number;

    constructor(values?: Partial<AssayUploadResultModel>) {
        Object.assign(this, values);
    }
}

export class AssayStateModel {
    [immerable] = true;

    readonly definitions: AssayDefinitionModel[];
    readonly definitionsError: string;
    readonly definitionsLoadingState: LoadingState;
    readonly protocolError: string;
    readonly protocolLoadingState: LoadingState;

    constructor(values?: Partial<AssayStateModel>) {
        Object.assign(this, values);

        this.definitions = this.definitions ?? [];
        this.definitionsLoadingState = this.definitionsLoadingState ?? LoadingState.INITIALIZED;
        this.protocolLoadingState = this.protocolLoadingState ?? LoadingState.INITIALIZED;
    }

    getById(assayRowId: number): AssayDefinitionModel {
        return this.definitions.find(def => def.id === assayRowId);
    }

    getByName(assayName: string): AssayDefinitionModel {
        const lowerName = assayName.toLowerCase();
        return this.definitions.find(def => def.name.toLowerCase() === lowerName);
    }

    getDefinitionsByTypes(included?: string[], excluded?: string[]): AssayDefinitionModel[] {
        if (!included && !excluded) return this.definitions;

        const lowerIncluded = included?.join('|').toLowerCase().split('|');
        const lowerExcluded = excluded?.join('|').toLowerCase().split('|');

        return this.definitions.filter(def => {
            let include = true;
            if (included?.length > 0) {
                include = lowerIncluded.indexOf(def.type.toLowerCase()) !== -1;
            }
            if (excluded?.length > 0) {
                include = lowerExcluded.indexOf(def.type.toLowerCase()) === -1;
            }
            return include;
        });
    }

    mutate(props: Partial<AssayStateModel>): AssayStateModel {
        return produce<AssayStateModel>(this, draft => {
            Object.assign(draft, props);
        });
    }
}
