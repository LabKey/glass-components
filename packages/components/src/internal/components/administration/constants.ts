import { PermissionRoles } from '@labkey/api';
import { Map } from 'immutable';

const ROLE_ACTION_DESCRIPTION =
    'and creating and editing sample types, source types, assay designs, and job templates.';

export const SECURITY_ROLE_DESCRIPTIONS = Map<string, string>([
    [
        PermissionRoles.ApplicationAdmin,
        'Application Administrators have full control over the application. This includes user management, permission assignments, ' +
            ROLE_ACTION_DESCRIPTION,
    ],
    [
        PermissionRoles.ProjectAdmin,
        'Project Administrators have full control over the application in this project container. This includes user management, permission assignments, ' +
            ROLE_ACTION_DESCRIPTION,
    ],
    [
        PermissionRoles.FolderAdmin,
        'Folder Administrators have control over the folder specific pieces of the application. This includes permission assignments, ' +
            ROLE_ACTION_DESCRIPTION,
    ],
    [
        PermissionRoles.Editor,
        'Editors may add new information or edit data related to samples, sources, assays, jobs, but not storage locations.',
    ],
    [PermissionRoles.Reader, 'Readers have a read-only view of the application.'],
]);

export const HOSTED_APPLICATION_SECURITY_ROLES = Map<string, string>([
    [PermissionRoles.Editor, 'Editor'],
    [PermissionRoles.EditorWithoutDelete, 'Editor without Delete'],
    [PermissionRoles.Reader, 'Reader'],
]);

export const SITE_SECURITY_ROLES = Map<string, string>([
    [PermissionRoles.ApplicationAdmin, 'Application Administrator'],
]);

export const APPLICATION_SECURITY_ROLES = Map<string, string>([
    [PermissionRoles.ProjectAdmin, 'Project Administrator'],
    [PermissionRoles.FolderAdmin, 'Folder Administrator'],
    [PermissionRoles.Editor, 'Editor'],
    [PermissionRoles.EditorWithoutDelete, 'Editor without Delete'],
    [PermissionRoles.Reader, 'Reader'],
]);


