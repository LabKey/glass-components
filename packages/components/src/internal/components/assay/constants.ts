import { SCHEMAS } from '../../schemas';

export const GENERAL_ASSAY_PROVIDER_NAME = 'General';

export const PLATE_METADATA_COLUMN = 'PlateMetadata';
export const PLATE_TEMPLATE_COLUMN = 'PlateTemplate';

// Columns are required for us to render the WorkflowTask in EditableDetails components
export const WORKFLOW_TASK_PROPERTIES_REQUIRED_COLUMNS = ['WorkflowTask', 'WorkflowTask/Run', 'Protocol/RowId'];

export const RUN_PROPERTIES_REQUIRED_COLUMNS = SCHEMAS.CBMB.concat(
    'Name',
    'RowId',
    'ReplacesRun',
    'ReplacedByRun',
    'DataOutputs',
    'DataOutputs/DataFileUrl',
    'Batch',
    'Folder'
)
    .concat(WORKFLOW_TASK_PROPERTIES_REQUIRED_COLUMNS)
    .toList();
