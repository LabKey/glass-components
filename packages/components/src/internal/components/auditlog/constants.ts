import { Query } from '@labkey/api';

export type AuditQuery = {
    containerFilter?: Query.ContainerFilter;
    hasDetail?: boolean;
    label: string;
    value: string;
};

export const ATTACHMENT_AUDIT_QUERY: AuditQuery = { value: 'attachmentauditevent', label: 'Attachment Events' };
export const DOMAIN_AUDIT_QUERY: AuditQuery = { value: 'domainauditevent', label: 'Domain Events' };
export const DOMAIN_PROPERTY_AUDIT_QUERY: AuditQuery = {
    value: 'domainpropertyauditevent',
    label: 'Domain Property Events',
};
export const DATA_UPDATE_AUDIT_QUERY: AuditQuery = {
    value: 'queryupdateauditevent',
    label: 'Data Update Events',
    hasDetail: true,
};
export const INVENTORY_AUDIT_QUERY: AuditQuery = {
    value: 'inventoryauditevent',
    label: 'Storage Management Events',
    hasDetail: true,
};
export const LIST_AUDIT_QUERY: AuditQuery = { value: 'listauditevent', label: 'List Events' };
export const GROUP_AUDIT_QUERY: AuditQuery = {
    value: 'groupauditevent',
    label: 'Roles and Assignment Events',
    containerFilter: Query.ContainerFilter.allFolders,
};
export const PROJECT_AUDIT_QUERY: AuditQuery = { value: 'containerauditevent', label: 'Project Events' };
export const SAMPLE_TYPE_AUDIT_QUERY: AuditQuery = { value: 'samplesetauditevent', label: 'Sample Type Events' };
export const SAMPLE_TIMELINE_AUDIT_QUERY: AuditQuery = {
    value: 'sampletimelineevent',
    label: 'Sample Timeline Events',
    hasDetail: true,
};
export const USER_AUDIT_QUERY: AuditQuery = {
    value: 'userauditevent',
    label: 'User Events',
    containerFilter: Query.ContainerFilter.allFolders,
};
export const ASSAY_AUDIT_QUERY: AuditQuery = { value: 'experimentauditevent', label: 'Assay Events' };
export const WORKFLOW_AUDIT_QUERY: AuditQuery = {
    value: 'samplesworkflowauditevent',
    label: 'Sample Workflow Events',
    hasDetail: true,
};
export const SOURCE_AUDIT_QUERY: AuditQuery = { value: 'sourcesauditevent', label: 'Sources Events', hasDetail: true };
export const COMMON_AUDIT_QUERIES: AuditQuery[] = [
    ATTACHMENT_AUDIT_QUERY,
    DOMAIN_AUDIT_QUERY,
    DOMAIN_PROPERTY_AUDIT_QUERY,
    DATA_UPDATE_AUDIT_QUERY,
    INVENTORY_AUDIT_QUERY,
    LIST_AUDIT_QUERY,
    GROUP_AUDIT_QUERY,
    SAMPLE_TYPE_AUDIT_QUERY,
    SAMPLE_TIMELINE_AUDIT_QUERY,
    USER_AUDIT_QUERY,
];

export const NOTEBOOK_AUDIT_QUERY: AuditQuery = {
    value: 'LabBookEvent',
    label: 'Notebook Events',
};

export const NOTEBOOK_REVIEW_AUDIT_QUERY: AuditQuery = {
    value: 'NotebookEvent',
    label: 'Notebook Review Events',
};

export const REGISTRY_AUDIT_QUERY: AuditQuery = { value: 'RegistryEvent', label: 'Registry Events' };

export const AUDIT_EVENT_TYPE_PARAM = 'eventType';
