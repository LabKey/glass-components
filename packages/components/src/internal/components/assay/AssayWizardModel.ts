import { ReactNode } from 'react';
import { List, Map, OrderedMap, Record } from 'immutable';
import { AssayDOM } from '@labkey/api';

import {
    AppURL,
    AssayDefinitionModel,
    FileAttachmentFormModel,
    getEditorModel,
    QueryColumn,
    QueryGridModel,
    QueryInfo,
} from '../../..';
import { AssayUploadTabs } from '../../constants';
import { generateNameWithTimestamp } from '../../util/Date';

// exported for jest testing
export function parseDataTextToRunRows(rawData: string): any[] {
    if (!rawData || !rawData.length) {
        return null;
    }

    const rows = [];
    let columns = [];

    rawData
        .split('\n')
        .filter(row => row.trim().length > 0)
        .forEach(row => {
            const parts = row.split('\t');
            if (parts.length === 0) return;

            if (columns.length === 0) columns = parts;
            else {
                const row = {};
                parts.forEach((part, index) => {
                    if (part.trim() !== '') {
                        row[columns[index]] = part;
                    }
                });
                rows.push(row);
            }
        });

    return rows.length > 0 ? rows : null;
}

export interface IAssayUploadOptions extends AssayDOM.IImportRunOptions {
    dataRows?: any; // Array<any> | QueryGridModel
    maxFileSize?: number;
    maxRowCount?: number;
}

export class AssayWizardModel
    extends Record({
        assayDef: undefined,
        isError: undefined,
        isWarning: false,
        isInit: false,
        isLoading: false,
        batchId: undefined, // batchId is null for first run
        lastRunId: undefined,
        returnUrl: undefined,
        isSubmitted: undefined,
        isSubmitting: undefined,
        errorMsg: undefined,

        attachedFiles: Map<string, File>(),
        batchColumns: OrderedMap<string, QueryColumn>(),
        batchProperties: Map<string, any>(),
        usePreviousRunFile: false,
        runColumns: OrderedMap<string, QueryColumn>(),
        runId: undefined,
        runProperties: Map<string, any>(),
        runName: undefined,
        comment: undefined,
        dataText: undefined,
        queryInfo: undefined,
        toDelete: undefined,
        selectedSamples: undefined,

        jobDescription: undefined,
        jobNotificationProvider: undefined,
        forceAsync: false,
    })
    implements FileAttachmentFormModel {
    assayDef: AssayDefinitionModel;
    isError?: boolean;
    isInit: boolean;
    isLoading: boolean;
    isWarning?: boolean;
    batchId?: number;
    lastRunId?: number;
    returnUrl?: AppURL;
    isSubmitted?: boolean;
    isSubmitting?: boolean;
    errorMsg?: ReactNode;

    attachedFiles: Map<string, File>;
    batchColumns: OrderedMap<string, QueryColumn>;
    batchProperties: Map<string, any>;
    usePreviousRunFile: boolean;
    runColumns: OrderedMap<string, QueryColumn>;
    runId?: string;
    runProperties?: Map<string, any>;
    runName?: string;
    comment?: string;
    dataText: string;
    queryInfo: QueryInfo;
    toDelete?: string;
    selectedSamples?: Map<string, any>;

    jobDescription?: string;
    jobNotificationProvider?: string;
    forceAsync?: boolean;

    isFilesTab(currentStep: AssayUploadTabs): boolean {
        return currentStep === AssayUploadTabs.Files;
    }

    isCopyTab(currentStep: AssayUploadTabs): boolean {
        return currentStep === AssayUploadTabs.Copy;
    }

    isGridTab(currentStep: AssayUploadTabs): boolean {
        return currentStep === AssayUploadTabs.Grid;
    }

    getAttachedFiles(): List<File> {
        return this.attachedFiles.valueSeq().toList();
    }

    getRunName(currentStep: AssayUploadTabs): string {
        if (this.runName) {
            return this.runName;
        }

        if (this.isFilesTab(currentStep)) {
            // Issue 39328: set assay run to 'undefined' so that the server will fill it in based on the file name
            // note that in the "file already exists so it will be renamed" case, that renamed file name will be used.
            const file = this.getAttachedFiles().first();
            if (file || this.usePreviousRunFile) {
                return undefined;
            }
        }

        return generateNameWithTimestamp(this.assayDef.name);
    }

    hasData(currentStep: number, gridModel: QueryGridModel): boolean {
        if (this.isFilesTab(currentStep)) {
            if (!this.attachedFiles.isEmpty()) {
                return true;
            }
            if (this.runId && this.usePreviousRunFile) {
                return true;
            }
        } else if (this.isCopyTab(currentStep)) {
            return this.dataText !== undefined;
        } else if (this.isGridTab(currentStep)) {
            // TODO add a dirty flag to the editorModel
            // const editorModel = getEditorModel(gridModel.getId());
            // return editorModel.hasData();
            return true;
        }
        return false;
    }

    prepareFormData(currentStep: number, gridModel: QueryGridModel): IAssayUploadOptions {
        const {
            batchId,
            batchProperties,
            comment,
            dataText,
            assayDef,
            runProperties,
            runId,
            usePreviousRunFile,
            jobDescription,
            jobNotificationProvider,
            forceAsync
        } = this;

        const assayData: any = {
            assayId: assayDef.id,
            batchId,
            batchProperties: batchProperties.toObject(),
            comment,
            name: this.getRunName(currentStep),
            properties: runProperties.toObject(),
            reRunId: runId,
            saveDataAsFile: true,
            jobDescription,
            jobNotificationProvider,
            forceAsync
        };

        Object.keys(assayData).forEach(k => {
            if (assayData[k] === undefined) {
                delete assayData[k];
            }
        });

        if (this.isFilesTab(currentStep)) {
            assayData.files = this.getAttachedFiles().toArray();
            if (runId !== undefined && usePreviousRunFile && assayData.files.length === 0) {
                const url = runProperties.get('DataOutputs/DataFileUrl');
                if (url) {
                    const filesIndex = url.indexOf('@files');
                    // get past the @files and the trailing slash
                    assayData.runFilePath = url.substring(filesIndex + 7);
                }
            }
        } else if (this.isCopyTab(currentStep)) {
            assayData.dataRows = parseDataTextToRunRows(dataText);
        } else if (this.isGridTab(currentStep)) {
            // need to get the EditorModel for the data to use in the import
            const editorModel = getEditorModel(gridModel.getId());

            assayData.dataRows = editorModel
                .getRawData(gridModel)
                .valueSeq()
                .map(row => row.filter(v => v !== undefined && v !== null && ('' + v).trim() !== ''))
                .toList()
                .toJS();
        } else {
            throw new Error('Unsupported upload step! Current step: "' + currentStep + '"');
        }

        // remove the "DataOutputs/DataFileUrl" from properties as it causes failures in AssayRunUploadContextImpl.addVocabularyProperties()
        if (assayData.properties['DataOutputs/DataFileUrl']) {
            delete assayData.properties['DataOutputs/DataFileUrl'];
        }

        return assayData;
    }
}
