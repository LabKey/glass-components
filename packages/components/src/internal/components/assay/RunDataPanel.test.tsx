import React from 'react';

import { act } from '@testing-library/react';

import { ASSAY_WIZARD_MODEL } from '../../../test/data/constants';
import { AssayUploadTabs } from '../../constants';
import { EditorModel } from '../editable/models';
import { TEST_USER_EDITOR } from '../../userFixtures';

import { withFormSteps, WithFormStepsProps } from '../forms/FormStep';
import { QueryModel } from '../../../public/QueryModel/QueryModel';
import { LoadingState } from '../../../public/LoadingState';

import { renderWithAppContext } from '../../test/reactTestLibraryHelpers';
import { Operation } from '../../../public/QueryColumn';

import { AssayDefinitionModel, AssayDomainTypes } from '../../AssayDefinitionModel';
import { QueryInfo } from '../../../public/QueryInfo';
import { TEST_LIMS_STARTER_MODULE_CONTEXT } from '../../productFixtures';

import { AssayWizardModel } from './AssayWizardModel';
import { RunDataPanel } from './RunDataPanel';

const MODEL_ID_NOT_LOADED = 'not loaded';
const MODEL_ID_LOADED = 'loaded';

interface OwnProps {
    showTabs?: boolean;
    wizardModel: AssayWizardModel;
}
type Props = OwnProps & WithFormStepsProps;

class RunDataPanelWrapperImpl extends React.Component<Props, any> {
    render() {
        const { currentStep, showTabs, wizardModel } = this.props;
        const queryInfo = wizardModel.queryInfo;
        const queryModel = new QueryModel({
            id: MODEL_ID_LOADED,
            schemaQuery: queryInfo.schemaQuery,
        }).mutate({
            rows: {},
            orderedRows: [],
            rowsLoadingState: LoadingState.LOADED,
            queryInfoLoadingState: LoadingState.LOADED,
            queryInfo,
        });
        const editorModel = new EditorModel({ id: MODEL_ID_LOADED });

        return (
            <RunDataPanel
                currentStep={currentStep}
                editorModel={editorModel}
                onDataFileChange={jest.fn()}
                onDataFileRemoval={jest.fn()}
                onGridChange={jest.fn()}
                onResultsFileChange={jest.fn()}
                onResultsFileRemoval={jest.fn()}
                onTextChange={jest.fn()}
                operation={Operation.insert}
                queryModel={queryModel}
                showTabs={showTabs}
                wizardModel={wizardModel}
            />
        );
    }
}

const RunDataPanelWrapper = withFormSteps(RunDataPanelWrapperImpl, {
    currentStep: AssayUploadTabs.Files,
    furthestStep: AssayUploadTabs.Files,
    hasDependentSteps: false,
});

describe('RunDataPanel', () => {
    test('loading state based on gridModel', () => {
        const queryModel = new QueryModel({
            id: MODEL_ID_LOADED,
            schemaQuery: ASSAY_WIZARD_MODEL.queryInfo.schemaQuery,
        });
        const editorModel = new EditorModel({ id: MODEL_ID_NOT_LOADED });
        const component = (
            <RunDataPanel
                currentStep={AssayUploadTabs.Files}
                editorModel={editorModel}
                onDataFileChange={jest.fn()}
                onDataFileRemoval={jest.fn()}
                onGridChange={jest.fn()}
                onResultsFileChange={jest.fn()}
                onResultsFileRemoval={jest.fn()}
                onTextChange={jest.fn()}
                operation={Operation.insert}
                queryModel={queryModel}
                wizardModel={ASSAY_WIZARD_MODEL}
            />
        );
        renderWithAppContext(component, { serverContext: { user: TEST_USER_EDITOR } });

        expect(document.querySelector('.panel-heading').textContent).toBe('Results');
        expect(document.querySelectorAll('.fa-spinner')).toHaveLength(1);
    });

    test('default props', () => {
        renderWithAppContext(<RunDataPanelWrapper wizardModel={ASSAY_WIZARD_MODEL} />, {
            serverContext: { user: TEST_USER_EDITOR },
        });
        expect(document.querySelector('.panel-heading').textContent).toBe('Results');
        expect(document.querySelectorAll('.fa-spinner')).toHaveLength(0);

        expect(document.querySelectorAll('.list-group')).toHaveLength(1);
        const tabs = document.querySelectorAll('.form-step-tab');
        expect(tabs).toHaveLength(2);
        expect(tabs[0].textContent).toBe('Enter Data into Grid');
        expect(tabs[1].textContent).toBe('Import Data from File');

        expect(document.querySelectorAll('.assay-data-file-label')).toHaveLength(0);

        expect(document.querySelectorAll('.file-upload--input')).toHaveLength(1);
        expect(document.querySelectorAll('.file-upload--file-entry-listing')).toHaveLength(1);
        expect(document.querySelectorAll('input[name="fileUpload"]')).toHaveLength(1);
        expect(document.querySelectorAll('input[name="fileUpload1"]')).toHaveLength(0);
    });

    test('showsTabs false', () => {
        renderWithAppContext(<RunDataPanelWrapper showTabs={false} wizardModel={ASSAY_WIZARD_MODEL} />, {
            serverContext: { user: TEST_USER_EDITOR },
        });
        expect(document.querySelectorAll('.list-group')).toHaveLength(0);
        expect(document.querySelectorAll('.form-step-tab')).toHaveLength(0);
    });

    test('assay model with results File field types', async () => {
        LABKEY.moduleContext = TEST_LIMS_STARTER_MODULE_CONTEXT;

        const modelWithFiles = AssayDefinitionModel.create({
            domains: {
                'GPAT 1 Batch Fields': [],
                'GPAT 1 Run Fields': [],
                'GPAT 1 Data Fields': [
                    { inputType: 'file', fieldKey: 'f1', name: 'f1' },
                    { inputType: 'text', fieldKey: 'other', name: 'other' },
                    { inputType: 'file', fieldKey: 'f2', name: 'f2' },
                ],
            },
            domainTypes: {
                Run: 'GPAT 1 Run Fields',
                Batch: 'GPAT 1 Batch Fields',
                Result: 'GPAT 1 Data Fields',
            },
        });
        const queryInfo = QueryInfo.fromJsonForTests({
            schemaName: 'assay',
            name: 'Data',
            columns: [
                { inputType: 'file', fieldKey: 'f1', name: 'f1' },
                { inputType: 'text', fieldKey: 'other', name: 'other' },
                { inputType: 'file', fieldKey: 'f2', name: 'f2' },
            ],
        });
        const wizardModel = new AssayWizardModel({
            isInit: true,
            assayDef: modelWithFiles,
            batchColumns: modelWithFiles.getDomainColumns(AssayDomainTypes.BATCH),
            runColumns: modelWithFiles.getDomainColumns(AssayDomainTypes.RUN),
            queryInfo,
        });

        let container;
        await act(async () => {
            container = renderWithAppContext(<RunDataPanelWrapper wizardModel={wizardModel} />, {
                serverContext: {
                    user: TEST_USER_EDITOR,
                },
            }).container;
        });
        expect(container).toMatchSnapshot();

        const fileDropLabels = document.querySelectorAll('.assay-data-file-label');
        expect(fileDropLabels).toHaveLength(2);
        expect(fileDropLabels[0].textContent).toBe('Results Data ');
        expect(fileDropLabels[1].textContent).toBe('File Fields Data ');

        expect(document.querySelectorAll('.file-upload--input')).toHaveLength(2);
        expect(document.querySelectorAll('.file-upload--file-entry-listing')).toHaveLength(2);
        expect(document.querySelectorAll('input[name="fileUpload"]')).toHaveLength(1);
        expect(document.querySelectorAll('input[name="fileUpload1"]')).toHaveLength(1);
    });
});
