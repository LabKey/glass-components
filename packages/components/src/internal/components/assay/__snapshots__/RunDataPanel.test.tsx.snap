// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunDataPanel assay model with results File field types 1`] = `
<div>
  <div
    class="panel panel-default"
  >
    <div
      class="panel-heading"
    >
      Results
    </div>
    <div
      class="panel-body"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12"
        >
          <ul
            class="list-group clearfix"
            style="list-style: none;"
          >
            <li
              class="list-group-item form-step-tab"
            >
              Enter Data into Grid
            </li>
            <li
              class="list-group-item form-step-tab active"
            >
              Import Data from File
            </li>
          </ul>
          <div
            class="clearfix"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-sm-12"
        >
          <div
            class="form-step"
          >
            <div>
              <div
                class="row editable-grid-buttons"
              >
                <div
                  class="col-sm-3"
                >
                  <div
                    class="editable-grid__controls text-nowrap"
                  >
                    <span
                      class="input-group input-group-align"
                    >
                      <input
                        class="form-control"
                        max="1000"
                        min="1"
                        name="addCount"
                        style="width: 65px;"
                        type="number"
                        value="1"
                      />
                      <span
                        class="input-group-btn"
                      >
                        <button
                          class="btn btn-primary"
                          type="button"
                        >
                          Add row
                        </button>
                      </span>
                    </span>
                  </div>
                </div>
                <div
                  class="editable-grid-buttons__action-buttons col-sm-9"
                >
                  <span
                    class="pull-right"
                  >
                    <div
                      class="export-menu"
                    >
                      <div
                        class="overlay-trigger"
                      >
                        <div
                          class="lk-dropdown btn-group dropdown"
                        >
                          <button
                            aria-expanded="false"
                            aria-haspopup="true"
                            class="btn btn-default dropdown-toggle"
                            id="dropdown-button-5"
                            role="button"
                            type="button"
                          >
                            <span
                              class="fa fa-download"
                            />
                          </button>
                          <ul
                            aria-labelledby="dropdown-button-5"
                            class="dropdown-menu dropdown-menu-right"
                            role="menu"
                          >
                            <li
                              class="lk-dropdown-header dropdown-header"
                              role="heading"
                            >
                              Export Data
                            </li>
                            <li
                              class="lk-menu-item"
                              role="presentation"
                            >
                              <a
                                href="#"
                                role="menuitem"
                              >
                                <div
                                  class="export-menu__item"
                                >
                                  <span
                                    class="fa fa-file-o export-menu-icon"
                                  />
                                  <span>
                                    CSV
                                  </span>
                                </div>
                              </a>
                            </li>
                            <li
                              class="lk-menu-item"
                              role="presentation"
                            >
                              <a
                                href="#"
                                role="menuitem"
                              >
                                <div
                                  class="export-menu__item"
                                >
                                  <span
                                    class="fa fa-file-excel-o export-menu-icon"
                                  />
                                  <span>
                                    Excel
                                  </span>
                                </div>
                              </a>
                            </li>
                            <li
                              class="lk-menu-item"
                              role="presentation"
                            >
                              <a
                                href="#"
                                role="menuitem"
                              >
                                <div
                                  class="export-menu__item"
                                >
                                  <span
                                    class="fa fa-file-text-o export-menu-icon"
                                  />
                                  <span>
                                    TSV
                                  </span>
                                </div>
                              </a>
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </span>
                </div>
              </div>
              <div
                class="editable-grid__container"
              >
                <div
                  class=""
                >
                  <div
                    class="grid-messages"
                  />
                  <table
                    class="table-cellular table-striped table-bordered"
                  >
                    <thead>
                      <tr>
                        <th
                          class="grid-header-cell"
                          draggable="false"
                          id="__editing__"
                          style="min-width: 45px;"
                        >
                          Row
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr
                        class="grid-empty"
                      >
                        <td
                          colspan="1"
                        >
                          Start by adding the quantity of assay data rows you want to create.
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
          <div
            class="form-step active"
          >
            <span
              class="translator--toggle__wizard"
            >
              <div
                class="row"
              >
                <div
                  class="col-sm-12"
                >
                  <div
                    class="assay-data-file-label"
                  >
                    <span>
                      Results Data
                      Â 
                      <div
                        class="overlay-trigger"
                      >
                        <i
                          class="fa fa-question-circle"
                        />
                      </div>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col-sm-12"
                >
                  <div
                    class="wizard-row--container"
                  >
                    <div
                      class=""
                    >
                      <div
                        class="file-upload--container block"
                      >
                        <label
                          class="file-upload--label"
                          for="fileUpload"
                        >
                          <i
                            aria-hidden="true"
                            class="fa fa-cloud-upload fa-2x cloud-logo"
                          />
                          Select file or drag and drop here.
                        </label>
                        <input
                          accept=""
                          class="file-upload--input"
                          id="fileUpload"
                          name="fileUpload"
                          type="file"
                        />
                      </div>
                      <div
                        class="file-upload--file-entry-listing"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </span>
            <div
              class="top-spacing"
            >
              <span
                class="translator--toggle__wizard"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-sm-12"
                  >
                    <div
                      class="assay-data-file-label"
                    >
                      <span>
                        File Fields Data
                        Â 
                        <div
                          class="overlay-trigger"
                        >
                          <i
                            class="fa fa-question-circle"
                          />
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="col-sm-12"
                  >
                    <div
                      class="wizard-row--container"
                    >
                      <div
                        class=""
                      >
                        <div
                          class="file-upload--container block"
                        >
                          <label
                            class="file-upload--label"
                            for="fileUpload1"
                          >
                            <i
                              aria-hidden="true"
                              class="fa fa-cloud-upload fa-2x cloud-logo"
                            />
                            Select file(s) or drag and drop here
                          </label>
                          <input
                            accept=""
                            class="file-upload--input"
                            id="fileUpload1"
                            multiple=""
                            name="fileUpload1"
                            type="file"
                          />
                        </div>
                        <div
                          class="file-upload--file-entry-listing"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="top-spacing"
      />
    </div>
  </div>
</div>
`;
